### DISCUSSION TITLE
initrd --timeout did not work?

### DISCUSSION THREAD

**Discussion by petrasl1976** [ipxe/ipxe#335] [Q&A] [answered]
Hi,

After network is setup:
```
set net0/ip $IP
set net0/netmask $MASK
set net0/gateway $GW
ifopen net0 && goto boot || goto reboot
```
I want to fetch images to boot from. Some providers has different network settings (STP delay on switch ports or smth), so actual network for iPXE become available after some time. Personally I saw up to 30s. I do not want to use "synthetic" `sleep 30` each time, so i expect `--timeout` for `initrd` command should do this trick:
```
initrd --timeout 64000 https://some.company.com/initrd || goto reboot
```
But even i set 64 seconds it still timeouts after about 12s. Probably i should use some kind of retry instead:
```
initrd --timeout 64000 https://some.company.com/initrd || initrd --timeout 64000 https://some.company.com/initrd || initrd --timeout 64000 https://some.company.com/initrd || goto reboot
```
Is there more proper way to wait for network become available before fetching images?

Thanks

**Comment by mcb30** [comment, marked as answer]
The `dhcp` command will automatically wait for delays caused by blocked links (e.g. STP delay, LACP delay, or 802.1x MAC bypass timeout).

There is unfortunately no viable way to wait for this class of blocked links without attempting to send some packets.  For some mechanisms (e.g. 802.1x), the delay timer won't even start ticking unless we attempt to send something.  There is no way to query the network to determine if a link is blocked: we know that a link is blocked only if we happen to receive something that indicates that the link is currently blocked, such as an STP packet with the forwarding flag unset.  Upon receiving such a packet, we know that the link should be considered as blocked for some defined period of time thereafter.

iPXE maintains a notion of a blocked link (which you can observe via "ifstat"), but this needs to be coupled with some mechanism such as DHCP that already allows for retries and timeouts in order to provide the overall behaviour of "extend timeouts as needed to allow for blocked links".  iPXE will extend the DHCP timeout up to `DHCP_DISC_MAX_DEFERRALS` (currently 3 minutes) in the presence of a link that is detected to be blocked.

The closest you can probably get with a static IP address is to use the `ping` command to check for connectivity to some known external IP address.  You can't directly detect a blocked link this way, but you can at least identify when external connectivity is available.

The simplest solution is almost certainly to use DHCP instead of static IP configuration.

**Reply by petrasl1976** [reply]
DHCP is not the way for my case. I have 1 or 2 servers per provider, so to setup extra server on their premises for DHCP or DHCP relay there wouldn't be cost efficient. Probably for now i will stick with retry option while downloading the firs image. I use ping check during initrd phase to wait for link, before get squashfs:
```
...
att=10
while [ ${att} -gt 0 ]
do
  ping -c 1 $pxeserver && break || echo "$pxeserver is not responding yet"
  let att=att-1; echo "attempts left $att"
done
log_begin_msg "Trying wget ${url} -O ${dest}/$(basename ${url})"
wget "${url}" -O "${dest}/$(basename ${url})"
...
```

Thanks for clarification

Regards
